### PR TITLE
fix(trace-handler): fix fold and canon compatibility

### LINKED ISSUE #356: Bug: fold and canon can lead to an invalid state merging
Due to current values stream merging scheme it's possible to merge states in inappropriate way if `canon` is used inside `fold`.

For example, given the annotated AIR script:

```
(seq
    (seq
        (call "relay" ("kad" "neighborhood") ["relay"] neighs_top) ; ok = ["p1"]
        (seq
            (call "p1" ("kad" "neighborhood") ["p1"] neighs_inner) ; ok =["p1"]
            (par
                (call "relay" ("peer" "identify") ["relay"] $external_addresses) ; behaviour = echo
                (call "p1" ("peer" "identify") ["p1"] $external_addresses) ; behaviour = echo
            )
        )
    )
    (seq
        (new $monotonic_stream
            (fold $external_addresses elem
                (seq
                    (ap "asd" $monotonic_stream)
                    (canon "relay" $monotonic_stream #result)
                )
            )
        )
        (call "client" ("return" "") [$external_addresses neighs_inner] x) ; ok = null
    )
)
```
when executed cyclically on the nodes `"client"`, `"relay"` and `"p1"`, eventually gives an outcome with an internal error:
```
RawAVMOutcome { 
   ret_code: 20000,
   error_message: "on instruction 'canon \"relay\" $monotonic_stream #result' trace handler encountered an error: position 6 from canon result CanonResult { stream_elements_pos: [6, 8] } hasn't been met yet",
   data: [...],
   call_requests: {},
   next_peer_pks: []
}
```

It happens because on merging previous and current data, unmet calls from the current data are all put into the same generation in the new data without particular order, and their ordering doesn't match the execution order.

### PR DESCRIPTION
**Pull request by mikevoronov**
Closes #356.

### CODE REVIEW

File: air/src/execution_step/boxed_value/stream.rs [hunk 1]
```diff
@@ -34,35 +38,51 @@ pub struct Stream {
     /// obtained values from a current_data that were not present in prev_data becomes a new generation.
     values: Vec<Vec<ValueAggregate>>,
 
+    /// Count of values from previous data.
+    previous_values_count: usize,
```
Comment (ValeryAntopol) [inline air/src/execution_step/boxed_value/stream.rs:42]: Maybe generations, not values?

File: air/src/execution_step/boxed_value/stream.rs [hunk 2]
```diff
@@ -34,35 +38,51 @@ pub struct Stream {
...
+    pub(crate) fn add_value(
+        &mut self,
+        value: ValueAggregate,
+        generation: Generation,
+        source: ValueSource,
+    ) -> ExecutionResult<u32> {
+        let generation = match (generation, source) {
+            (Generation::Last, _) => self.values.len() - 1,
+            (Generation::Nth(id), ValueSource::PreviousData) => id as usize,
+            (Generation::Nth(id), ValueSource::CurrentData) => self.previous_values_count + id as usize,
```
Comment (ValeryAntopol) [inline air/src/execution_step/boxed_value/stream.rs:85]: Dont like word `id` here. `index` looks better. Or `original_generation`.

File: air/src/execution_step/boxed_value/stream.rs [hunk 3]
```diff
@@ -182,6 +202,26 @@ impl Stream {
 
         Some(iter)
     }
+
+    /// Removes empty generations updating data and returns final generation count.
+    pub(crate) fn compactify(mut self, trace_ctx: &mut TraceHandler) -> ExecutionResult<usize> {
```
Comment (ValeryAntopol) [inline air/src/execution_step/boxed_value/stream.rs:207]: Why not `&mut self`? To make sure it is the last call to the structure?

File: air/src/farewell_step/outcome.rs [hunk 4]
```diff
@@ -104,12 +112,23 @@ fn populate_outcome_from_contexts(
...
-        error_message,
-        data,
-        next_peer_pks,
-        call_requests,
+    InterpreterOutcome::new(ret_code, error_message, data, next_peer_pks, call_requests)
+}
+
+// this method is called only if there is an internal error in the interpreter and
+// new execution trace was corrupted
+fn execution_error_into_outcome(
```
Comment (ValeryAntopol) [inline air/src/farewell_step/outcome.rs:120]: This comment is wrong. This method is called always and it decides if it was an error or not. I suggest something like this in the calling code:
```match maybe_gens.map_err(execution_error_into_outcome)```
Then this comment will decribe reality properly.

File: air/tests/test_module/features/streams/streams_early_exit.rs [hunk 5]
```diff
@@ -226,6 +226,7 @@ fn fold_early_exit() {
         "",
         last_error_receiver_result.data
     );
+    print_trace(&last_peer_checker_result, "");
```
Comment (ValeryAntopol) [inline air/tests/test_module/features/streams/streams_early_exit.rs:229]: Forgotten debug log

File: air/tests/test_module/features/streams/streams_early_exit.rs [hunk 6]
```diff
@@ -342,6 +343,7 @@ fn fold_par_early_exit() {
...
```
Comment (ValeryAntopol) [inline air/tests/test_module/features/streams/streams_early_exit.rs:346]: ditto

File: crates/air-lib/test-utils/src/lib.rs [hunk 7]
```diff
@@ -132,5 +132,12 @@ pub fn is_interpreter_succeded(result: &RawAVMOutcome) -> bool {
 }
 
 pub fn check_error(result: &RawAVMOutcome, error: impl ToErrorCode + ToString) -> bool {
+    println!(
```
Comment (ValeryAntopol) [inline crates/air-lib/test-utils/src/lib.rs:135]: ditto

File: air/src/execution_step/boxed_value/stream.rs (hunk 3 above)
Comment (mikevoronov) [reply air/src/execution_step/boxed_value/stream.rs:207]: Yep, to provide additional hardening

File: air/src/execution_step/boxed_value/stream.rs [hunk 8]
```diff
@@ -34,35 +38,51 @@ pub struct Stream {
...
     /// This map is intended to support canonicalized stream creation, such streams has
     /// corresponding value positions in a data and this field are used to create such streams.
     values_by_pos: HashMap<TracePos, StreamValueLocation>,
 }
 
 impl Stream {
-    pub(crate) fn from_generations_count(count: usize) -> Self {
+    pub(crate) fn from_generations_count(previous_count: usize, current_count: usize) -> Self {
+        let last_generation_count = 1;
+        // TODO: check for overflow
```
Comment (monoid) [inline air/src/execution_step/boxed_value/stream.rs:52]: I think it should be implemented before the tomorrow event.  Even if it just panics.

File: air/src/execution_step/execution_context/streams_variables.rs [hunk 9]
```diff
@@ -33,36 +40,32 @@ pub(crate) struct Streams {
...
+    /// should have at the scope start.
+    previous_restricted_stream_gens: RestrictedStreamGens,
+
+    /// Contains stream generations from current data that a restricted stream
+    /// should have at the scope start.
+    current_restricted_stream_gens: RestrictedStreamGens,
 
     /// Contains stream generations that each private stream had at the scope end.
     /// Then it's placed into data
     collected_restricted_stream_gens: RestrictedStreamGens,
```
Comment (monoid) [inline air/src/execution_step/execution_context/streams_variables.rs:53]: Worth renaming to `new_restricted_stream_gens` for consistency.

**Review by monoid — APPROVED** [review]

**Pull request merged**